src/Host/Client/Impl/Session/RSession.cs

```diff
@@ -317,26 +309,27 @@ internal sealed class RSession : IRSession, IRCallbacks {
             Disconnected?.Invoke(this, EventArgs.Empty);
 
             var currentRequest = Interlocked.Exchange(ref _currentRequestSource, null);
-            currentRequest?.CompleteResponse();
+            var exception = new RHostDisconnectedException();
+            currentRequest?.TryCancel(exception);
 
-            ClearPendingRequests();
+            ClearPendingRequests(exception);
 
             return Task.CompletedTask;
         }
 
-        private void ClearPendingRequests() {
+        private void ClearPendingRequests(OperationCanceledException exception) {
             RSessionRequestSource requestSource;
```

Comment:
This method is called from 3 places:
- `CancelAllAsync` - should cancel all currently available requests, but it doesn't block any new requests from being added, so we don't need it to be atomic in this case
- `StartHostAsyncBackground` - no new requests can be added when method is called because RHost is not running yet
- `IRCallbacks.Disconnected` - RHost is already stopped, so no new requests can be added

---
In reply to: [71067562](https://github.com/Microsoft/RTVS/pull/2080#discussion_r71067562) [](ancestors = 71067562)